File: 20_ValidParentheses/20_v1.cpp
class Solution {
public:
    bool isValid(string s) {
        map<char, char> mapping;
        mapping[')'] = '(';
        mapping['}'] = '{';
        mapping[']'] = '[';
        
        stack<char> check;
        
        for (char c: s){
            map<char, char>::iterator it(mapping.find(c));
            if (it != mapping.end()){
                if (check.empty() || it->second != check.top())
                    return false;
                check.pop();
            } else {
                check.push(c);
            }
        }
        
        return check.empty();
    }
};

File: 20_ValidParentheses/20_v1.js
/**
 * @param {string} s
 * @return {boolean}
 */
var isValid = function(s) {
    var mapping = {
        ')': '(',
        ']': '[',
        '}': '{',
    }
    
    check = []
    
    for (let i = 0; i < s.length; i++){
        var c = s[i]
        if (mapping[c] !== undefined){
            if (check.length === 0 || check.slice(-1)[0] != mapping[c])
                return false;
            check.pop()
        } else {
            check.push(c)
        }
    }
    
    return check.length === 0
};
